odoo/fields.py

```diff
@@ -3010,15 +3010,32 @@ def validated(domain):
     def _description_domain(self, env):
         domain = self.domain(env[self.model_name]) if callable(self.domain) else self.domain  # pylint: disable=not-callable
         if self.check_company:
-            # when using check_company=True on a field on 'res.company', the
-            # company_id comes from the id of the current record
+            field_to_check = None
             if self.company_dependent:
-                cid = 'allowed_company_ids[0]'
+                # when using check_company=True on a field on 'res.company', the
+                # company_id comes from the id of the current record
+                cids = '[allowed_company_ids[0]]'
+            elif self.model_name == 'res.company':
+                cids = '[id]'
+            elif 'company_id' in env[self.model_name]:
+                cids = '[company_id]'
+                field_to_check = 'company_id'
+            elif 'company_ids' in env[self.model_name]:
+                cids = 'company_ids'
+                field_to_check = 'company_ids'
             else:
-                cid = "id" if self.model_name == "res.company" else "company_id"
-            company_domain = env[self.comodel_name]._check_company_domain(companies=unquote(cid))
-            no_company_domain = env[self.comodel_name]._check_company_domain(companies='')
-            return f"({cid} and {company_domain} or {no_company_domain}) + ({domain or []})"
+                _logger.warning(_(
+                    "Couldn't generate a company-dependent domain for field %s. "
+                    "The model doesn't have a 'company_id' or 'company_ids' field, and isn't company-dependent either.",
+                    f'{self.model_name}.{self.name}'
+                ))
+                return domain
+            company_domain = env[self.comodel_name]._check_company_domain(companies=unquote(cids))
+            if not field_to_check:
```

Comment:
You can't do it like this, this depends on the value interpreted in the form view